File: src/main/fetcher.ts
import { getConfig } from "./store";
import { fetchRepo, hasRemote } from "./git";

// MARK: Periodic background fetch

/** How often to `git fetch` every configured repo: 8 minutes and 43 seconds. */
export const FETCH_INTERVAL_MS = (8 * 60 + 43) * 1000;

let timer: ReturnType<typeof setInterval> | null = null;
let running = false;

/**
 * Fetch every configured repo that has a remote, in parallel. Never throws:
 * one repo's failure (offline, auth prompt suppressed, etc.) can't abort the
 * others or the loop. Returns the number of repos that fetched successfully.
 */
async function fetchAllRepos(): Promise<number> {
  const { repos } = getConfig();
  const results = await Promise.allSettled(
    repos.map(async (repo) => {
      if (!(await hasRemote(repo.path))) return false;
      await fetchRepo(repo.path);
      return true;
    }),
  );

  let fetched = 0;
  for (let i = 0; i < results.length; i++) {
    const r = results[i];
    if (r.status === "fulfilled") {
      if (r.value) fetched++;
    } else {
      console.warn(`auto-fetch: ${repos[i]?.name ?? "repo"} failed:`, r.reason);
    }
  }
  return fetched;
}

/**
 * Run one fetch cycle now, then every FETCH_INTERVAL_MS. `onFetched` fires after
 * any cycle that fetched at least one repo, so the renderer can refresh
 * ahead/behind and unpushed counts against the newly updated remote refs.
 * Overlapping cycles are skipped (a slow network can't stack fetches).
 */
export function startAutoFetch(onFetched: () => void): void {
  if (timer) return;

  const cycle = async (): Promise<void> => {
    if (running) return;
    running = true;
    try {
      if ((await fetchAllRepos()) > 0) onFetched();
    } finally {
      running = false;
    }
  };

  void cycle();
  timer = setInterval(() => void cycle(), FETCH_INTERVAL_MS);
}

/** Stop the background fetch loop (idempotent). */
export function stopAutoFetch(): void {
  if (timer) clearInterval(timer);
  timer = null;
}

File: crates/wtm-core/src/fetcher.rs
//! Periodic background `git fetch` of every configured repo.

use std::time::Duration;

use log::warn;

use crate::git::{fetch_repo, has_remote};
use crate::types::RepoConfig;

/// How often to fetch: 8 minutes and 43 seconds.
pub const FETCH_INTERVAL: Duration = Duration::from_secs(8 * 60 + 43);

/// Fetch every repo that has a remote, concurrently. Never fails; returns how
/// many repos fetched successfully.
pub async fn fetch_all(repos: Vec<RepoConfig>) -> usize {
    let mut set = tokio::task::JoinSet::new();
    for repo in repos {
        set.spawn(async move {
            if !has_remote(&repo.path, "origin").await {
                return false;
            }
            match fetch_repo(&repo.path).await {
                Ok(()) => true,
                Err(e) => {
                    warn!("auto-fetch: {} failed: {e}", repo.name);
                    false
                }
            }
        });
    }
    let mut fetched = 0;
    while let Some(r) = set.join_next().await {
        if r.unwrap_or(false) {
            fetched += 1;
        }
    }
    fetched
}
